src/TesApi.Web/BatchPoolService.cs

```diff
@@ -0,0 +1,118 @@
+        private readonly ILogger _logger;
+        private readonly bool _isDisabled;
+
+        /// <summary>
+        /// Interval between each call to <see cref="IBatchPool.ServicePoolAsync(CancellationToken)"/>.
+        /// </summary>
+        public static readonly TimeSpan RunInterval = TimeSpan.FromSeconds(30);
+
+        /// <summary>
+        /// Default constructor
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <param name="batchScheduler"></param>
+        /// <param name="logger"></param>
+        /// <exception cref="ArgumentNullException"></exception>
+        public BatchPoolService(IConfiguration configuration, IBatchScheduler batchScheduler, ILogger<BatchPoolService> logger)
+        {
+            _batchScheduler = batchScheduler ?? throw new ArgumentNullException(nameof(batchScheduler));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _isDisabled = configuration.GetValue("BatchAutopool", false);
+        }
+
+        /// <inheritdoc />
+        public override Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (_isDisabled)
+            {
+                return Task.CompletedTask;
+            }
+
+            return base.StartAsync(cancellationToken);
+        }
+
+        /// <inheritdoc />
+        public override Task StopAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Scheduler stopping...");
+            return base.StopAsync(cancellationToken);
+        }
+
+        /// <inheritdoc />
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            _logger.LogInformation("Scheduler started.");
+
+            while (!stoppingToken.IsCancellationRequested)
+            {
+                try
+                {
+                    await ServiceBatchPools(stoppingToken);
+                    await Task.Delay(RunInterval, stoppingToken);
+                }
+                catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
+                {
+                    break;
+                }
+                catch (Exception exc)
+                {
+                    _logger.LogError(exc, exc.Message);
+                }
+            }
+
+            _logger.LogInformation("Scheduler gracefully stopped.");
+        }
+
+        /// <summary>
+        /// Retrieves all batch pools from the database and affords an opportunity to react to changes.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async ValueTask ServiceBatchPools(CancellationToken cancellationToken)
+        {
+            var pools = _batchScheduler.GetPools().ToList();
+
+            if (0 == pools.Count)
+            {
+                return;
+            }
+
+            var startTime = DateTime.UtcNow;
+
+            foreach (var pool in pools)
```

Comment:
Would it be better to service multiple pools simultaneously?  Is API throttling the main concern?

Reply:
Processing each pool was a tiny fraction of the time needed to process each TesTask when manually adjusting pool sizes, and by dropping all that work by utilizing auto-scaling, it's now even less. Processing them simultaneously is more likely to add throttling than it is to save time.